libs/search-widget/src/core/old-stores/main.store.ts

```diff
@@ -60,6 +61,74 @@ export const nucliaStore = (): NucliaStore => {
         map((results) => getSortedResources(results)),
         startWith([] as IResource[]),
       ),
+      smartResults: _store.searchResults.pipe(
+        filter((res) => !!res.resources),
+        map((results) => {
+          const allResources = results.resources;
+          if (!allResources || Object.keys(allResources).length === 0) {
+            return [] as Search.SmartResult[];
+          }
+          const fullTextResults =
+            results.fulltext?.results.map((res) => allResources[res.rid]).filter((res) => !!res) || [];
+          const semanticResults = results.sentences?.results || [];
+          let smartResults: Search.SmartResult[] = [];
+
+          // best fulltext match goes first
+          if (fullTextResults.length > 0) {
+            smartResults.push(fullTextResults[0]);
+          }
+
+          // if not a keyword search, add the 2 best semantic sentences
+          const looksLikeKeywordSearch = _store!.query.value.split(' ').length < 3;
+          if (!looksLikeKeywordSearch) {
+            const twoBestSemantic = semanticResults.slice(0, 2);
+            twoBestSemantic.forEach((sentence) => {
+              const resource = allResources[sentence.rid];
+              if (resource) {
+                smartResults = addParagraphToSmartResults(
+                  smartResults,
+                  resource,
+                  generateFakeParagraphForSentence(allResources, sentence),
+                );
+              }
+            });
+          }
+
+          // add the rest of the fulltext results
+          const existingResourceIds = smartResults.map((res) => res.id);
+          const remainingFullTextResults = fullTextResults
+            .slice(1)
```

Comment:
you tested `if (fullTextResults.length > 0)` before adding the first fulltext result to `smartResults`, should you run this test here as well before `fullTextResults.slice(1)`?
I feel like nothing prevent you to arrive here if `fullTextResults.length === 0` (even though I feel like that would be a super corner case)